### DISCUSSION TITLE
i3 Window Cycler, an Alt+Tab alternative to i3-cycle-focus.py

### DISCUSSION THREAD

**Discussion by arakis** [i3/i3#6238] [Show and tell]
## Introduction to i3 Window Cycler

I've developed alternative to `i3-cycle-focus.py`: https://github.com/arakis/i3-window-cycler

It allows you to switch between windows based on their recent usage without a visual popup. It emulates the Alt+Tab functionality found in other desktop environments by cycling window focus in a most-recently-used (MRU) order.

### Improvements over i3-cycle-focus.py
    
While `i3-cycle-focus.py` uses a timeout to detect when the user has finished switching windows, this method can cause issues when you're switching windows quickly. The script may mistakenly think you're still in switching mode, leading to unintended behavior.

### Why a Key Listener Is Required

The i3 window manager does not natively support binding actions to the release of modifier keys (like Alt) using the `--release` flag in its configuration. To overcome this limitation, the i3 Window Cycler uses a separate key listener script that monitors keyboard events.

Importantly, this [key listener](https://github.com/arakis/i3-window-cycler/blob/main/i3-key-listener.py) captures only the single modifier key release event (e.g., when the Alt key is released) and does **not** send the keycode to any process.

### Future Improvements
When [this](https://github.com/arakis/i3-window-cycler/blob/93c3ad40ca65aeb08229c0a155a15c28a6201548/README.md?plain=1#L96) is supported in the future, we will no longer need the key-listener

**Comment by parisni** [comment]
Hi, interesting solution for the main i3 limitation: window navigation.

You might be interested in the similar tool I just released https://github.com/parisni/dfzf

Found inspiration in your autofocus and for sure in the release action you did !
